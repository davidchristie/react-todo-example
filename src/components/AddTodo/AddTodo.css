.AddTodo {
  padding: 5px;
}

.AddTodo input {
  font-size: larger;
  padding: 10px;
  width: 80%;
}
