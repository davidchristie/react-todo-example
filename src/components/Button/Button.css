.Button {
  background: white;
  border: none;
  box-shadow: 0px 0px 3px black;
  color: black;
  font-size: larger;
  padding: 10px;
}
