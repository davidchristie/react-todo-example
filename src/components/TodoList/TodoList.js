import PropTypes from 'prop-types'
import React from 'react'

import TodoItem from '../TodoItem'

const TodoList = props => {
  return (
    <div className='TodoList'>
      {
        props.items.map((item, index) => {
          return (
            <TodoItem
              item={item}
              key={index}
              onClick={() => props.toggleCompleted(index)}
            />
          )
        })
      }
    </div>
  )
}

TodoList.propTypes = {
  items: PropTypes.arrayOf(PropTypes.object).isRequired,
  toggleCompleted: PropTypes.func.isRequired
}

export default TodoList
