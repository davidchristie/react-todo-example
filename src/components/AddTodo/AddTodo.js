import PropTypes from 'prop-types'
import React from 'react'

import './AddTodo.css'

class AddTodo extends React.Component {
  constructor (props) {
    super(props)
    this._onChange = this._onChange.bind(this)
    this._onSubmit = this._onSubmit.bind(this)
    this.state = {
      text: ''
    }
  }

  _isValidTodo () {
    return this.state.text.length > 0
  }

  _onChange (event) {
    this.setState({
      text: event.target.value
    })
  }

  _onSubmit (event) {
    event.preventDefault()
    if (this._isValidTodo()) {
      if (this.props.onSubmit) {
        const newItem = {
          completed: false,
          text: this.state.text
        }
        this.props.onSubmit(newItem)
      }
      this.setState({
        text: ''
      })
    }
  }

  render () {
    return (
      <div className='AddTodo'>
        <form onSubmit={this._onSubmit}>
          <input
            onChange={this._onChange}
            placeholder='Add an item...'
            type='text'
            value={this.state.text}
          />
        </form>
      </div>
    )
  }
}

AddTodo.propTypes = {
  onSubmit: PropTypes.func
}

export default AddTodo
