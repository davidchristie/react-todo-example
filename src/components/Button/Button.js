import PropTypes from 'prop-types'
import React from 'react'

import './Button.css'

const Button = props => {
  const { className, ...attributes } = props
  let combinedClassName = 'Button'
  if (className) {
    combinedClassName += ' ' + className
  }
  return (
    <button
      className={combinedClassName}
      {...attributes}
    />
  )
}

Button.propTypes = {
  className: PropTypes.string
}

export default Button
