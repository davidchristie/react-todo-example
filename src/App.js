import React from 'react'

import AddTodo from './components/AddTodo'
import Button from './components/Button'
import Header from './components/Header'
import TodoList from './components/TodoList'
import './App.css'

const defaultItems = [
  {
    completed: false,
    text: 'Install Node.js version >= 6'
  },
  {
    completed: false,
    text: 'Install npm'
  },
  {
    completed: false,
    text: 'Install Yarn'
  },
  {
    completed: false,
    text: 'Use npm to globally install create-react-app'
  }
]

class App extends React.Component {
  constructor (props) {
    super(props)
    this._addTodo = this._addTodo.bind(this)
    this._deleteCompleted = this._deleteCompleted.bind(this)
    this._toggleCompleted = this._toggleCompleted.bind(this)
    this.state = {
      items: defaultItems
    }
  }

  _addTodo (newItem) {
    this.setState({
      items: [
        ...this.state.items,
        newItem
      ]
    })
  }

  _deleteCompleted () {
    this.setState({
      items: this.state.items.filter(item => {
        return !item.completed
      })
    })
  }

  _toggleCompleted (itemIndex) {
    this.setState({
      items: this.state.items.map((item, index) => {
        if (index === itemIndex) {
          return {
            ...item,
            completed: !item.completed
          }
        } else {
          return item
        }
      })
    })
  }

  render () {
    const hasCompleted = this.state.items.some(item => {
      return item.completed
    })
    return (
      <div className='App'>
        <Header />
        <TodoList
          items={this.state.items}
          toggleCompleted={this._toggleCompleted}
        />
        <AddTodo onSubmit={this._addTodo} />
        {
          hasCompleted
            ? (
              <Button onClick={this._deleteCompleted}>
                Remove Completed
              </Button>
            )
            : null
        }
      </div>
    )
  }
}

export default App
