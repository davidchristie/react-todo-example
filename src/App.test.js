import { shallow } from 'enzyme'
import React from 'react'

import AddTodo from './components/AddTodo'
import Button from './components/Button'
import TodoList from './components/TodoList'
import App from './App'

it('renders correctly', () => {
  expect(shallow(<App />)).toMatchSnapshot()
})

it('toggles item completion', () => {
  const state = {
    items: [
      {
        completed: false,
        text: 'Test item 1'
      },
      {
        completed: false,
        text: 'Test item 2'
      },
      {
        completed: false,
        text: 'Test item 3'
      }
    ]
  }
  const itemIndex = 1
  const wrapper = shallow(<App />)
  wrapper.setState(state)
  wrapper.find(TodoList).at(0).props().toggleCompleted(itemIndex)
  expect(wrapper.state().items[itemIndex].completed).toBe(!state.items[itemIndex].completed)
})

it('adds a new todo item', () => {
  const state = {
    items: [
      {
        completed: false,
        text: 'Test item 1'
      },
      {
        completed: false,
        text: 'Test item 2'
      },
      {
        completed: false,
        text: 'Test item 3'
      }
    ]
  }
  const newItem = {
    completed: false,
    text: 'Test item 4'
  }
  const wrapper = shallow(<App />)
  wrapper.setState(state)
  wrapper.find(AddTodo).at(0).props().onSubmit(newItem)
  expect(wrapper.state().items[3]).toBe(newItem)
})

it('removes completed todo items', () => {
  const state = {
    items: [
      {
        completed: true,
        text: 'Test item 1'
      },
      {
        completed: false,
        text: 'Test item 2'
      },
      {
        completed: true,
        text: 'Test item 3'
      }
    ]
  }
  const wrapper = shallow(<App />)
  wrapper.setState(state)
  wrapper.find(Button).at(0).simulate('click')
  expect(wrapper.state().items).toEqual(state.items.filter(item => !item.completed))
})
