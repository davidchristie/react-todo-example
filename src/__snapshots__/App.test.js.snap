// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="App"
>
  <Header />
  <TodoList
    items={
      Array [
        Object {
          "completed": false,
          "text": "Install Node.js version >= 6",
        },
        Object {
          "completed": false,
          "text": "Install npm",
        },
        Object {
          "completed": false,
          "text": "Install Yarn",
        },
        Object {
          "completed": false,
          "text": "Use npm to globally install create-react-app",
        },
      ]
    }
    toggleCompleted={[Function]}
  />
  <AddTodo
    onSubmit={[Function]}
  />
</div>
`;
